llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -1096,6 +1096,53 @@ Value *InstCombinerImpl::foldUsingDistributiveLaws(BinaryOperator &I) {
   return SimplifySelectsFeedingBinaryOp(I, LHS, RHS);
 }
 
+std::optional<std::pair<Value *, Value *>>
+InstCombinerImpl::matchSymmetricPhiNodesPair(PHINode *LHS, PHINode *RHS) {
+
+  if (LHS->getParent() != RHS->getParent())
+    return std::nullopt;
+
+  if (LHS->getNumIncomingValues() < 2)
+    return std::nullopt;
+
+  BasicBlock *B0 = LHS->getIncomingBlock(0);
+  Value *N1 = LHS->getIncomingValueForBlock(B0);
```

Comment:
Using getIncomingValueForBlock() is inefficient for large phis (the function will be quadratic).

What you can do is add a check `if (!equal(LHS->blocks(), RHS->blocks()) return false;` and then just work on the incoming values ignoring the blocks.

InstCombine canonicalizes the block order for phis in the same block, so this check only guards against phi nodes that have no been canonicalized yet -- it will eventually succeed.

Reply:
Is there any efficient way to `equal(LHS->blocks(), RHS->blocks())` ?

Reply:
Doesn't writing literally that work?

Reply:
Oh, it works. Thanks a lot !